Modules/HTTPServer/Sources/HTTPServer/Server/Server.swift

```diff
@@ -157,36 +157,46 @@ extension Server {
     }
 
     public func process(stream: Stream) throws {
-        let parser = RequestParser(stream: stream, bufferSize: bufferSize)
+        let chunkBytes = UnsafeMutablePointer<UInt8>.allocate(capacity: 4096)
+        defer { chunkBytes.deallocate(capacity: 4096) }
+        let chunk = UnsafeMutableBufferPointer<UInt8>(start: chunkBytes, count: 4096)
+        
+        let parser = Parser(mode: .request)
         let serializer = ResponseSerializer(stream: stream, bufferSize: bufferSize)
-
+        
         while !stream.closed {
             do {
-                let request = try parser.parse()
-                let response = try middleware.chain(to: responder).respond(to: request)
-                try serializer.serialize(response)
-
-                if let upgrade = response.upgradeConnection {
-                    try upgrade(request, stream)
-                    stream.close()
-                }
-
-                if !request.isKeepAlive {
-                    stream.close()
+                let bytesRead = try stream.read(into: chunk)
+                let chunkRead = UnsafeBufferPointer(start: chunkBytes, count: bytesRead)
+                
+                try parser.parse(chunkRead) { message in
+                    let request = message as! Request
```

Comment:
It's a tiny bit of a hit, but it's only one typecast per fully parsed request. I don't think it's going to make any difference.